README.md
# tr-064

TR-064 - UPnP/IGD for node.js

## Description

A library to interact with routers and other network devices.
Tested and designd for Fritz.Box routers.

This library is capable of:

- Supports the UPnP, IGD and PMR (Samsung TV) Protocol
- Read and configure Services
- Authentication with username/password or password only
- SSL encryption
- Transactions
- Subscribe to Events with included EventServer

More info about TR-064: https://avm.de/fileadmin/user_upload/Global/Service/Schnittstellen/AVM_TR-064_first_steps.pdf

## Install

<pre>
  npm install @nicolaischmid/tr-064
</pre>

## It`s simple

Connect to the device and read a Service.

```javascript
var tr = require('tr-064');
var tr064 = new tr.TR064();
tr064.initTR064Device('fritz.box', 49000, function(err, device) {
  if (!err) {
    var wanip = device.services['urn:dslforum-org:service:WANIPConnection:1'];
    wanip.actions.GetInfo(function(err, result) {
      console.log(result);
    });
  }
});
```

## Save communication (SSL Encryption, Authentication)

```javascript
var tr = require('tr-064');
var tr064 = new tr.TR064();
tr064.initTR064Device('fritz.box', 49000, function(err, device) {
  if (!err) {
    device.startEncryptedCommunication(function(err, sslDev) {
      if (!err) {
        sslDev.login([USER], [PASSWORD]);
        var wanip = sslDev.services['urn:dslforum-org:service:WANIPConnection:1'];
        wanip.actions.GetInfo(function(err, result) {
          console.log(result);
        });
      }
    });
  }
});
```

## List All Services and Variables

Get the info from both protocols.

```javascript
var tr = require('tr-064');
var tr064 = new tr.TR064();
tr064.initTR064Device('fritz.box', 49000, function(err, device) {
  if (!err) {
    console.log('Found device! - TR-064');
    showDevice(device);
  }
});

tr064.initIGDDevice('fritz.box', 49000, function(err, device) {
  if (!err) {
    console.log('Found device! - IGD');
    showDevice(device);
  }
});

var showDevice = function(device) {
  console.log('=== ' + device.meta.friendlyName + ' ===');
  device.meta.servicesInfo.forEach(function(serviceType) {
    var service = device.services[serviceType];
    console.log('  ---> ' + service.meta.serviceType + ' <---');
    service.meta.actionsInfo.forEach(function(action) {
      console.log('   # ' + action.name + '()');
      action.inArgs.forEach(function(arg) {
        console.log('     IN : ' + arg);
      });
      action.outArgs.forEach(function(arg) {
        console.log('     OUT: ' + arg);
      });
    });
  });
};
```

## Methods

### initTR064Device(host, port, callback)

Initialize the TR - 064 UPnP controller

- `host` - hostname of the device
- `port` - port of the device(standard: 49000)
- `callback` - (err, device)

### initIGDDevice(host, port, callback)

Initialize the TR - 064 IGD controller

- `host` - hostname of the device
- `port` - port of the device(standard: 49000)
- `callback` - (err, device)

### device.startEncryptedCommunication([caFile],callback)

Starts SSL encrypted Communication

- `caFile` - Filename of custom .pem file (Optional)
- `callback` - (err, device)

### device.stopEncryptedCommunication()

Stops SSL encrypted Communication

### device.login([user],password)

Configure device to use authentication for every request

- `user` - Username (Optional, default device user is used instead)
- `password` - Device password

### device.logout()

Configure device to not use authentication

### device.startTransaction(callback)

Starts a 'device-side' transaction

- `callback` - (err, device)

### device.stopTransaction()

Ends the current transaction

### device.meta

Array with all info about services and actions

### device.services[`Service Identifier`]

Gets the specified service form the device

- `Service Identifier` - usually in the form of: urn:dslforum-org:service:XXX:1

### service.actions.XXX([{name: 'xx', value: 'xx'}], callback)

- `name` - Argumentname for Action
- `value` - Argumentvalue for Action
- `callback` - (err, result)

```javascript
service.actions.SetEnable([{ name: 'NewEnable', value: '1' }], function(err, result) {
  console.log(result);
});
```

### Contributing

Feel free to provide PRs with updates to this package. But please follow these steps:

- Fork it
- Add your changes
- Format your code: `yarn format`
- Commit your changes, while adhearing to the Angular commit convention.
- Submit your PR


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="0.4.5"></a>
## [0.4.5](https://github.com/NicolaiSchmid/tr064/compare/v0.4.4...v0.4.5) (2019-08-29)



<a name="0.4.4"></a>
## [0.4.4](https://github.com/NicolaiSchmid/tr064/compare/v0.4.3...v0.4.4) (2019-02-27)


### Bug Fixes

* import internal modules with esm ([3c72308](https://github.com/NicolaiSchmid/tr064/commit/3c72308))



<a name="0.4.3"></a>
## [0.4.3](https://github.com/NicolaiSchmid/tr064/compare/v0.4.2...v0.4.3) (2019-02-27)


### Bug Fixes

* actionRequestError ([1ab9288](https://github.com/NicolaiSchmid/tr064/commit/1ab9288)), closes [#1](https://github.com/NicolaiSchmid/tr064/issues/1)
* import device into into fix with esm ([caecaa1](https://github.com/NicolaiSchmid/tr064/commit/caecaa1))



<a name="0.4.2"></a>

## [0.4.2](https://github.com/NicolaiSchmid/tr064/compare/v0.4.1...v0.4.2) (2019-02-26)

### Bug Fixes

- if serviceType == wrong comparison ([a8574e4](https://github.com/NicolaiSchmid/tr064/commit/a8574e4))

<a name="0.4.1"></a>

## [0.4.1](https://github.com/NicolaiSchmid/tr064/compare/v0.4.0...v0.4.1) (2018-05-25)

<a name="0.4.0"></a>

# [0.4.0](https://github.com/NicolaiSchmid/tr064/compare/v0.3.2...v0.4.0) (2018-05-25)

### Features

- changes to service.js for better requests by @SeydX ([d6cf30c](https://github.com/NicolaiSchmid/tr064/commit/d6cf30c))

<a name="0.3.2"></a>

## [0.3.2](https://github.com/NicolaiSchmid/tr064/compare/v0.3.1...v0.3.2) (2018-05-25)

<a name="0.3.1"></a>

## 0.3.1 (2018-05-25)


test/index.js
const inspect = require('eyes').inspector({
  maxLength: false,
  hideFunctions: false,
  styles: {
    all: 'black',
  },
});
const util = require('util');
const tr = require('./lib/TR064');

const showCallback = function(err, result) {
  if (!err) {
    inspect(result);
  }
};

const speedCB = function(err, result) {
  util.print(`${result.NewByteReceiveRate  }  `);
};

const showDevice = function(device) {
  console.log(`=== ${  device.meta.friendlyName  } ===`);
  device.meta.servicesInfo.forEach((serviceType) => {
    const service = device.services[serviceType];
    console.log(`  ---> ${  service.meta.serviceType  } <---`);
    service.meta.actionsInfo.forEach((action) => {
      console.log(`   # ${  action.name  }()`);
      action.inArgs.forEach((arg) => {
        console.log(`     IN : ${  arg}`);
      });
      action.outArgs.forEach((arg) => {
        console.log(`     OUT: ${  arg}`);
      });
    });
  });
};

const tr064 = new tr.TR064();

// tr064.startEventServer(44880);

tr064.initIGDDevice('fritz.box', 49000, (err, device) => {
  if (!err) {
    console.log('Found device! - IGD');
    showDevice(device);

    const wandic = device.services['urn:schemas-upnp-org:service:WANIPConnection:1'];

    wandic.actions.GetExternalIPAddress(showCallback);

    const wan = device.services['urn:schemas-upnp-org:service:WANCommonInterfaceConfig:1'];
    wan.actions.GetCommonLinkProperties(showCallback);
    wan.actions.GetTotalBytesReceived(showCallback);
    wan.actions.GetAddonInfos(showCallback);
    setInterval(wan.actions.GetAddonInfos, 1000, speedCB);
  }
});

/*
Example: Get the url from the first phonebook.
*/
var user = 'user';
var password = 'password';
tr064.initTR064Device('fritz.box', 49000, (err, device) => {
  if (!err) {
    console.log('Found device! - TR-064');
    device.startEncryptedCommunication((err, sslDev) => {
      if (!err) {
        sslDev.login(user, password);
        const wanppp = sslDev.services['urn:dslforum-org:service:X_AVM-DE_OnTel:1'];
        wanppp.actions.GetPhonebook(
          {
            NewPhonebookID: '0',
          },
          (err, ret) => {
            if (err) {
              console.log(err);
              return;
            }
            if (ret.NewPhonebookURL && ret.NewPhonebookURL.length > 0) {
              // GOT URL
              const url = ret.NewPhonebookURL;
              console.log(url);
            }
          },
        );
      } else {
        console.log(err);
      }
    });
  }
});

/*
Example: Turn off/on WLAN on FritzBoxes.
*/
var user = 'user';
var password = 'password';
tr064.initTR064Device('fritz.box', 49000, (err, device) => {
  if (!err) {
    console.log('Found device! - TR-064');
    device.startEncryptedCommunication((err, sslDev) => {
      if (!err) {
        sslDev.login(user, password);
        const wlan24 = sslDev.services['urn:dslforum-org:service:WLANConfiguration:1'];
        const wlan5 = sslDev.services['urn:dslforum-org:service:WLANConfiguration:2'];
        const wlanGuest = sslDev.services['urn:dslforum-org:service:WLANConfiguration:3'];
        //                                                1 for on and 0 for off
        wlan24.actions.SetEnable([{ name: 'NewEnable', value: '1' }], (err, result) => {
          console.log(err);
          console.log(result);
        });
        //                                                1 for on and 0 for off
        wlan5.actions.SetEnable([{ name: 'NewEnable', value: '1' }], (err, result) => {
          console.log(err);
          console.log(result);
        });
        //                                                1 for on and 0 for off
        wlanGuest.actions.SetEnable([{ name: 'NewEnable', value: '1' }], (err, result) => {
          console.log(err);
          console.log(result);
        });
      } else {
        console.log(err);
      }
    });
  }
});


src/index.js
import Device from './Device';

const parseString = require('xml2js').parseString;
const request = require('request');
const http = require('http');
const inspect = require('eyes').inspector({
  maxLength: false,
  hideFunctions: false,
});
const URL = require('url');

const TR064_DESC_URL = '/tr64desc.xml';
const IGD_DESC_URL = '/igddesc.xml';
const PMR_DESC_URL = '/pmr/PersonalMessageReceiver.xml';

function TR064() {}
TR064.prototype.discoverDevice = function() {};
TR064.prototype.initTR064Device = function(host, port, callback) {
  this._parseDesc(host, port, TR064_DESC_URL, callback);
};
TR064.prototype.initIGDDevice = function(host, port, callback) {
  this._parseDesc(host, port, IGD_DESC_URL, callback);
};
TR064.prototype.initPMRDevice = function(host, port, callback) {
  this._parseDesc(host, port, PMR_DESC_URL, callback);
};

TR064.prototype.startEventServer = function(port) {
  this.eventServer = http.createServer((req, res) => {
    inspect(req);
    res.writeHead(200);
    res.end();
  });
  this.eventServer.listen(port);
};

TR064.prototype.stopEventServer = function() {
  this.removeAllEvents();
  this.server.close();
};

TR064.prototype.removeAllEvents = function() {};

TR064.prototype._parseDesc = function(host, port, url, callback) {
  const nurl = `http://${  host  }:${  port  }${url}`;
  request(nurl, (error, response, body) => {
    if (!error && response.statusCode == 200) {
      parseString(body, { explicitArray: false }, (err, result) => {
        if (!err) {
          // this.deviceInfo.push(result);
          const devInfo = result.root.device;
          devInfo.host = host;
          devInfo.port = port;
          const path = URL.parse(nurl).pathname;
          devInfo.urlPart = path.substring(0, path.lastIndexOf('/'));
          new Device(devInfo, callback);
        } else {
          console.log(err);
          console.log(result);
          callback(err, null);
        }
      });
    } else {
      console.log(error);
      console.log(body);
      callback(error, null);
    }
  });
};

exports.TR064 = TR064;


src/Device.js
import Service from './Service';

const inspect = require('eyes').inspector({
  maxLength: false,
  hideFunctions: false,
});
const async = require('async');
const request = require('request');
const crypto = require('crypto');

export default function Device(deviceInfo, callback) {
  this.meta = deviceInfo;
  this.meta.servicesInfo = [];
  this.readyCallback = callback;
  this.services = {};
  this._isTransaction = false;
  this._parseServices();
  this._sslPort = null;
  this._auth = {
    uid: null,
    realm: 'F!Box SOAP-Auth',
    chCount: 0,
  };
}
Device.prototype.listServices = function() {};
Device.prototype.listStateVariables = function() {};
Device.prototype.login = function(user, password) {
  if (password === undefined) {
    this._auth.uid = 'DefaultUser';
    this._auth.pwd = user;
  } else {
    this._auth.uid = user;
    this._auth.pwd = password;
  }
};
Device.prototype.logout = function() {
  this._auth.uid = null;
  this._auth.pwd = null;
  this._auth.chCount = 0;
};
Device.prototype.startTransaction = function(cb) {
  const that = this;
  const sessionID = uuid();
  this._startTransaction(sessionID, err => {
    if (!err) {
      that._isTransaction = true;
      cb(null, that);
    } else {
      cb(err, null);
    }
  });
};
Device.prototype.stopTransaction = function(cb) {
  const that = this;
  this._stopTransaction(err => {
    if (!err) {
      that._isTransaction = false;
      cb(null, that);
    } else {
      cb(err, null);
    }
  });
};
Device.prototype.startEncryptedCommunication = function(caFile, cb) {
  if (typeof caFile === 'function') {
    cb = caFile;
    caFile = null;
  } else {
    this._ca = fs.readFileSync(caFile); // .pem File
  }
  const that = this;
  this._getSSLPort((err, port) => {
    if (!err) {
      that._sslPort = port;
      cb(null, that);
    } else {
      cb(err, null);
    }
  });
};
Device.prototype.stopEncryptedCommunication = function() {
  this._sslPort = null;
};

var getServicesFromDevice = function(serviceArray, device) {
  serviceArray = serviceArray.concat(device.serviceList.service);
  // console.log(serviceArray);
  if (device.deviceList && Array.isArray(device.deviceList.device)) {
    device.deviceList.device.forEach(dev => {
      serviceArray = getServicesFromDevice(serviceArray, dev);
    });
  } else if (device.deviceList && device.deviceList.device) {
    serviceArray = getServicesFromDevice(serviceArray, device.deviceList.device);
  }
  return serviceArray;
};

Device.prototype._parseServices = function() {
  const serviceArray = getServicesFromDevice([], this.meta);
  const asyncAddService = bind(this, this._addService);
  const asyncAddResultToServiceList = bind(this, this._addResultToServiceList);
  async.concat(serviceArray, asyncAddService, asyncAddResultToServiceList);
};

Device.prototype._addService = function(serviceData, callback) {
  new Service(this, serviceData, callback);
};

Device.prototype._addResultToServiceList = function(err, services) {
  if (!err) {
    for (const i in services) {
      const service = services[i];
      this.services[service.meta.serviceType] = service;
      this.meta.servicesInfo.push(service.meta.serviceType);
    }
    delete this.meta.deviceList;
    delete this.meta.serviceList;
    this.readyCallback(null, this);
  } else {
    console.log(err);
    this.readyCallback(err, null);
  }
};

// SSL Encription
Device.prototype._getSSLPort = function(cb) {
  const devInfo = this.services['urn:dslforum-org:service:DeviceInfo:1'];
  devInfo.actions.GetSecurityPort((err, result) => {
    if (!err) {
      const sslPort = parseInt(result.NewSecurityPort);
      if (typeof sslPort === 'number' && isFinite(sslPort)) {
        cb(null, sslPort);
      } else {
        cb(new Error(`Got bad port from Device. Port:${result.NewSecurityPort}`));
      }
    } else {
      console.log(err);
      cb(new Error('Encription is not supported for this device.'));
    }
  });
};

// Login
Device.prototype._calcAuthDigest = function(uid, pwd, realm, sn) {
  let MD5 = crypto.createHash('md5');
  MD5.update(`${uid}:${realm}:${pwd}`);
  const secret = MD5.digest('hex');
  MD5 = crypto.createHash('md5');
  MD5.update(`${secret}:${sn}`);
  return MD5.digest('hex');
};

// Transaction
Device.prototype._startTransaction = function(sessionID, cb) {
  const devConfig = this.services['urn:dslforum-org:service:DeviceConfig:1'];
  devConfig.actions.ConfigurationStarted({ NewSessionID: sessionID }, err => {
    if (!err) {
      cb(null);
    } else {
      cb(new Error('Transactions are not supported for this device.'));
    }
  });
};
Device.prototype._stopTransaction = function(cb) {
  const devConfig = this.services['urn:dslforum-org:service:DeviceConfig:1'];
  devConfig.actions.ConfigurationFinished(err => {
    if (!err) {
      cb(null);
    } else {
      cb(new Error('Transactions are not supported for this device.'));
    }
  });
};

function uuid(
  a, // placeholder
) {
  return a // if the placeholder was passed, return
    ? // a random number from 0 to 15
      (
        a ^ // unless b is 8,
        ((Math.random() * // in which case
          16) >> // a random number from
          (a / 4))
      ) // 8 to 11
        .toString(16) // in hexadecimal
    : // or otherwise a concatenated string:
      (
        [1e7] + // 10000000 +
        -1e3 + // -1000 +
        -4e3 + // -4000 +
        -8e3 + // -80000000 +
        -1e11
      ) // -100000000000,
        .replace(
          // replacing
          /[018]/g, // zeroes, ones, and eights with
          uuid, // random hex digits
        );
}

function bind(scope, fn) {
  return function() {
    return fn.apply(scope, arguments);
  };
}


src/Service.js
const { parseString } = require('xml2js');
const inspect = require('eyes').inspector({
  maxLength: false,
  hideFunctions: false,
});
const request = require('request');

export default function Service(device, serviceInfo, callback) {
  this.host = device.meta.host;
  this.port = device.meta.port;
  this.device = device;
  this.meta = serviceInfo;
  this.meta.actionsInfo = [];
  this.readyCallback = callback;
  this.actions = {};
  this.stateVariables = {};
  this.logAttempts = [];
  // console.log("Service: "+this.host);
  _parseSCPD(this);
}

Service.prototype.listActions = function() {};
Service.prototype.listStateVariables = function() {};

const _pushArg = function(argument, inArgs, outArgs) {
  if (argument.direction === 'in') {
    inArgs.push(argument.name);
  } else if (argument.direction === 'out') {
    outArgs.push(argument.name);
  }
};

const _parseActions = function(actionData) {
  if (!Array.isArray(actionData)) {
    return;
  }
  const insA = bind(this, _insertAction);
  actionData.forEach(insA);
  // inspect(actions);
  //
};

const _parseSCPD = function(obj) {
  if (obj.device.meta.urlPart && obj.device.meta.urlPart.length > 0) {
    obj.meta.SCPDURL = `${obj.device.meta.urlPart}/${obj.meta.SCPDURL}`;
  }
  const url = `http://${obj.host}:${obj.port}${obj.meta.SCPDURL}`;
  // console.log(url);
  request(url, (error, response, body) => {
    if (!error && response.statusCode === 200) {
      // console.log(body);
      parseString(
        body,
        {
          explicitArray: false,
        },
        (err, result) => {
          const pA = bind(obj, _parseActions);
          const pV = bind(obj, _parseStateVariables);
          pA(result.scpd.actionList.action);
          pV(result.scpd.serviceStateTable.stateVariable);
          // inspect(obj.stateVariables);
          obj.readyCallback(null, obj);
        },
      );
    } else {
      console.log(url);
    }
  });
};

var _insertAction = function(el) {
  const outArgs = [];
  const inArgs = [];
  if (el.argumentList && Array.isArray(el.argumentList.argument)) {
    el.argumentList.argument.forEach(argument => {
      _pushArg(argument, inArgs, outArgs);
    });
  } else if (el.argumentList) {
    _pushArg(el.argumentList.argument, inArgs, outArgs);
  }

  this.actions[el.name] = bind(this, function(vars, callback) {
    this._callAction(el.name, inArgs, outArgs, vars, callback);
  });
  this.meta.actionsInfo.push({
    name: el.name,
    inArgs,
    outArgs,
  });
};

Service.prototype._callAction = function(name, inArguments, outArguments, vars, callback) {
  if (typeof vars === 'function') {
    callback = vars;
    vars = [];
  }

  bind(
    this,
    this._sendSOAPActionRequest(
      this.device,
      this.meta.controlURL,
      this.meta.serviceType,
      name,
      inArguments,
      outArguments,
      vars,
      callback,
    ),
  );
};

Service.prototype._subscribeStateVariableChangeEvent = function(sv, callback) {
  inspect(arguments);
};

function bind(scope, fn) {
  return function() {
    return fn.apply(scope, arguments);
  };
}

const _insertStateVariables = function(sv) {
  if (sv.$.sendEvents === 'yes') {
    this.stateVariables[sv.name] = bind(this, function(callback) {
      this._subscribeStateVariableChangeEvent(sv, callback);
    });
  }
};

const _parseStateVariables = function(stateVariableData) {
  const insSV = bind(this, _insertStateVariables);
  if (Array.isArray(stateVariableData)) {
    stateVariableData.forEach(insSV);
  } else if (typeof stateVariableData === 'object') {
    insSV(stateVariableData);
  }
};

Service.prototype._sendSOAPActionRequest = function(
  device,
  url,
  serviceType,
  action,
  inArguments,
  outArguments,
  vars,
  callback,
) {
  const self = this;
  let head = '';
  if (device._auth.uid) {
    // Content Level Authentication
    if (device._auth.auth) {
      head =
        `${'<s:Header>' +
          '<h:ClientAuth xmlns:h="http://soap-authentication.org/digest/2001/10/"' +
          's:mustUnderstand="1">' +
          '<Nonce>'}${device._auth.sn}</Nonce>` +
        `<Auth>${device._auth.auth}</Auth>` +
        `<UserID>${device._auth.uid}</UserID>` +
        `<Realm>${device._auth.realm}</Realm>` +
        `</h:ClientAuth>` +
        `</s:Header>`;
    } else {
      // First Auth
      head =
        `${' <s:Header>' +
          '<h:InitChallenge xmlns:h="http://soap-authentication.org/digest/2001/10/"' +
          's:mustUnderstand="1">' +
          '<UserID>'}${device._auth.uid}</UserID>` +
        `<Realm>${device._auth.realm}</Realm>` +
        `</h:InitChallenge>` +
        `</s:Header>`;
    }
  }

  let body =
    `${'<?xml version="1.0" encoding="utf-8"?>' +
      '<s:Envelope s:encodingStyle="http://schemas.xmlsoap.org/soap/encoding/" xmlns:s=" http://schemas.xmlsoap.org/soap/envelope/">'}${head}<s:Body>` +
    `<u:${action} xmlns:u="${serviceType}">`;

  Object.keys(vars).forEach(key => {
    body += `<${vars[key].name}>`;
    body += vars[key].value;
    body += `</${vars[key].name}>`;
  });

  body = `${body}</u:${action}>` + `</s:Body>` + `</s:Envelope>`;

  let port = 0;
  let proto = '';
  let agentOptions = null;

  if (device._sslPort) {
    port = device._sslPort;
    proto = 'https://';
    if (device._ca) {
      agentOptions = {
        ca: device._ca,
      };
    } else {
      agentOptions = {
        rejectUnauthorized: false,
      }; // Allow selfsignd Certs
    }
  } else {
    proto = 'http://';
    port = device.meta.port;
  }
  const uri = `${proto + device.meta.host}:${port}${url}`;
  const that = this;
  request(
    {
      method: 'POST',
      uri,
      agentOptions,
      headers: {
        SoapAction: `${serviceType}#${action}`,
        'Content-Type': 'text/xml; charset="utf-8"',
      },
      body,
      timeout: 5000,
    },
    (error, response, body) => {
      if (!error && response.statusCode === 200) {
        parseString(
          body,
          {
            explicitArray: false,
          },
          (err, result) => {
            let challange = false;
            let res = {};
            const env = result['s:Envelope'];
            if (env['s:Header']) {
              const header = env['s:Header'];
              if (header['h:Challenge']) {
                const ch = header['h:Challenge'];
                challange = true;
                if (self.logAttempts.length) {
                  for (const i in self.logAttempts) {
                    if (self.logAttempts[i].service === serviceType) {
                      if (self.logAttempts[i].attempts >= 1) {
                        error = new Error('Credentials incorrect');
                      } else {
                        self.logAttempts[i].attempts += 1;
                        device._auth.des = serviceType;
                        device._auth.sn = ch.Nonce;
                        device._auth.realm = ch.Realm;
                        device._auth.auth = device._calcAuthDigest(
                          device._auth.uid,
                          device._auth.pwd,
                          device._auth.realm,
                          device._auth.sn,
                        );
                        device._auth.chCount += 1;
                        that._sendSOAPActionRequest(
                          device,
                          url,
                          serviceType,
                          action,
                          inArguments,
                          outArguments,
                          vars,
                          callback,
                        );
                        return;
                      }
                    }
                  }
                } else {
                  self.logAttempts.push({ service: serviceType, attempts: 1 });
                  device._auth.sn = ch.Nonce;
                  device._auth.realm = ch.Realm;
                  device._auth.auth = device._calcAuthDigest(
                    device._auth.uid,
                    device._auth.pwd,
                    device._auth.realm,
                    device._auth.sn,
                  );
                  device._auth.chCount += 1;
                  // Repeat request.
                  that._sendSOAPActionRequest(
                    device,
                    url,
                    serviceType,
                    action,
                    inArguments,
                    outArguments,
                    vars,
                    callback,
                  );
                  return;
                }
              } else if (header['h:NextChallenge']) {
                const nx = header['h:NextChallenge'];
                for (const i in self.logAttempts) {
                  if (self.logAttempts[i].service === serviceType) {
                    self.logAttempts[i].attempts = 0;
                  }
                }
                // device._auth.auth = nx.Nonce;
                device._auth.chCount = 0;
                device._auth.sn = nx.Nonce;
                device._auth.realm = nx.Realm;
                device._auth.auth = device._calcAuthDigest(
                  device._auth.uid,
                  device._auth.pwd,
                  device._auth.realm,
                  device._auth.sn,
                );
              }
            }

            if (env['s:Body']) {
              const body = env['s:Body'];
              if (body[`u:${action}Response`]) {
                const responseVars = body[`u:${action}Response`];
                if (outArguments) {
                  outArguments.forEach(arg => {
                    res[arg] = responseVars[arg];
                  });
                }
              } else if (body['s:Fault']) {
                const fault = body['s:Fault'];
                error = new Error(`Device responded with fault ${error}`);
                res = fault;
              }
            }
            callback(error, res);
          },
        );
      } else {
        const newError = new Error(
          `sendSOAPActionRequest Error! [${action}] [${serviceType}]${
            response ? ` [${response.statusCode}]` : ''
          }${error ? ` [${error.code}]` : ''}`,
        );
        callback(newError, null);
      }
    },
  );
};

Service.prototype.sendSOAPEventSubscribeRequest = function(callback) {
  console.log('Send EventSubscribe...');
  request(
    {
      method: 'SUBSCRIBE',
      uri: `http://${this.host}:${this.port}${this.meta.eventSubURL}`,
      headers: {
        CALLBACK: '<http://192.168.178.28:44880/>',
        NT: 'upnp:event',
        TIMEOUT: 'Second-infinite',
      },
    },
    (error, response, body) => {
      console.log('END');
      if (response.statusCode === 200) {
        console.log('EventSubscribeRequest OK');
      } else {
        error = new Error(`EventSubscribeRequest Error: ${response.statusCode}`);
        console.log(`error: ${response.statusCode}`);
        console.log(body);
      }
    },
  );
};
